design/baremetal-operator/raid-api.md

```diff
@@ -0,0 +1,202 @@
+
+### Non-Goals
+
+1. Provide default RAID configuration.
+2. Retain existing RAID configuration on a host if it matches the
+   desired configuration at the time of provisioning (we make no
+   assumptions about preserving the state of the storage system or its
+   contents).
+3. Re-provision a host if the RAID settings change. (Users are
+   expected to explicitly deprovision the host and then reprovision
+   it.)
+4. Support every possible RAID configuration.
+
+## Proposal
+
+### User Stories
+
+1. As a user, I can specify the RAID configuration for a host before
+   provisioning it.
+2. As a user, I can choose between hardware and software RAID for each
+   host.
+
+## Design Details
+
+### API Changes
+
+The `BareMetalHost` API will change to accept RAID configuration
+inputs and to reflect the settings that were used when a host is
+provisioned (in case they are changed after provisioning).
+
+New `spec` fields will be added in a section called `raid`, with
+separate fields describing the hardware and software RAID settings. If
+the hardware settings are provided, the software settings are ignored.
+
+For hardware RAID, the inputs describe how to combine physical disks
+into a RAID volume via the RAID controller hardware in the host.
+
+```yaml
+spec:
+  raid:
+    hardwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        name: volume-name
+        rotational: true
+        numberOfPhysicalDisks: 2
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The number of disks to use for the volume defaults to a minimum
+  value based on the RAID level.
+
+For software RAID, the inputs describe how to configure the operating
+system to see physical disks in a RAID configuration through software
+drivers, without using any special hardware.
+
+```yaml
+spec:
+  raid:
+    softwareVolumes:
+      - sizeGibibytes: 1000
+        level: 1
+        physicalDisks:
+          - deviceName: /dev/sda
+          - deviceName: /dev/sdb
+```
+
+- The RAID level for a volume is required and should be expressed as
+  one of `"1"`, `"0"`, or `"1+0"`.
+- The size for a volume is optional and should be expressed in
+  gibibytes.
+- The disks to use are specified using the same syntax as the
+  `rootDeviceHints`.
+
+Software RAID is constrained in other ways:
+
+- There must be 0, 1, or 2 software RAID devices.
+- The first software RAID volume must use RAID level 1 to ensure the
+  host boots in the event of a disk failure.
+- The second software RAID volume may use RAID level `0`, `1`, or
+  `1+0`.
+
+*NOTE: Specifying the RAID level for the root volume seems like a
+recommendation we could make but not something we should really be
+requiring. Does something actually enforce that rule?*
+
+Regardless of the type of RAID used, the first volume defined is used
+as the root volume.
```

Comment:
How does this rule interact with the root device hints? What if a user specifies both?

Reply:
```
is_root_volume - Set to true if this is the root volume. At most one logical disk can have this set to true; the other logical disks must have this set to false. The root device hint will be saved, if the RAID interface is capable of retrieving it. This is false by default.
```

Reply:
Maybe @dtantsur, @iurygregory, or @elfosardo can help us understand the way root device hints and the root volume flag on the RAID setup interact.

Reply:
I'm sure since the ``is_root_volume`` is an optional flag (its default value being false), the following would make the most sense:
1. The first logical disk is root device of the configuration, if there is no ``is_root_device`` flag on any disk.
2. If the ``is_root_device`` is flag set on any logical disk, it would take precedence over statement 1.
3. If the ``is_root_device`` is flag is set on more than 1 logical disks, Ironic would complain, meaning the conductor will fail deployment with an error, moving Node to ``deploy_failed/clean_failed`` state.
4. The drivers can pull out this flag, to implement some wrapping functionality, if desired. Since the flag is saved in the Ironic database.

Reply:
The `idrac` hardware type/driver is not capable of retrieving the **root device hint**, so it does not save it.

Reply:
Oh I see. So we can scratch number 4 in case of iDRAC. It appears to be a stretch goal anyway, thanks for pointing it out @rpioso

Reply:
Instead of using a flag, could we say that the first item in the array will be set as the root volume and if RAID is used then any root device hints are ignored?

Reply:
I agree with the second part of the statement, it will make sense since RAID will be abstracting the hardware and presenting logical volumes for consumption as root volumes. 

About the first part of the statement, what's the significance of asserting it? Will it somehow simplify our explanation? Make the implementation more effective? I'm unclear. 
I assume we want to abstract away Ironic-land things, and in that regard, getting rid of an Ironic-land flag and using the BMO-land array references would help us in that endeavor? Is that what we're looking for?

Reply:
Yes, I'm looking for ways to make the metal3 API simple to understand. I don't want to over-simplify, though. I'm not sure why someone would, for example, define 3 RAID volumes and say the 2nd one should be the root volume for the OS. Is that a real use case? Is it a use case we care about for provisioning nodes?

Reply:
I agree. It's stretching it too far to presume it to be a real use case. I'm inclined to believe that to ignore the ``root_device_hints`` in case of RAID is good to go for us. I'd say what's in the implementation at this point should be good enough, unless the implementation processes the ``root_device_hints`` in case of RAID (which I doubt it does).

Reply:
`is_root_volume` makes ironic automatically overwrite root device hints with the created volume. If we don't use this flag, we need to set root device hints to point at the desired volume, otherwise ironic will apply its usual logic. In any case, I think "root volume is the first volume and must use RAID-1" is a sane thing to assert on our level. Will save us a lot of headache.

Reply:
I agree with @dtantsur. Let's stick to this.

Reply:
I think we're saying that metal3 should set the `is_root_volume` flag on the first volume and not give the user an option to set it themselves on other volumes. I will update this part of the doc with enough detail to make that clear.

Reply:
Makes sense. Let's make things easier for the user.

Reply:
The idea that the first RAID volume is always the root device, and it only uses RAID 1 provides a lot of convenience to the user. I just have one concern that what if we want the root device is a disk that is not a RAID disk? What if we only want to have RAID disks to store workload? How do you think @dhellmann @demonCoder95 @dtantsur ?

Reply:
For that we have what we call the [``root device hints``](https://github.com/metal3-io/baremetal-operator/blob/master/docs/api.md#rootdevicehints) which are a separate part of the BMH specification. 
If we only want RAID volumes for workload, I'm assuming we'll use ``root device hints`` to choose a physical disk as a root disk, and leave the rest up to RAID to consume to create RAID volumes for the workloads.

Reply:
@demonCoder95 but as discussed above, I can see that we have a plan to ignore `rootDeviceHint` if RAID volume is used. 
>Instead of using a flag, could we say that the first item in the array will be set as the root volume and if RAID is used then any root device hints are ignored?

For the use cased that I mentioned, I believe we should choose the first RAID volume as the root device when `rootDeviceHint` is not set. Otherwise, we should apply the hint to choose the root device.

Reply:
I think the discussion was not about ``rootDeviceHints`` as @zaneb mentioned earlier. It will make complete sense to me, if ``rootDeviceHints`` were used to override the default behavior of "always using first RAID volume as root disk".

Reply:
The next draft will say that the default is to use the first RAID volume as the root device unless rootDeviceHints are set.